### DISCUSSION TITLE
The optimal approach to define Prisma schema models and underlying database tables

### DISCUSSION THREAD

**Discussion by pavittarx** [prisma/prisma#2532] [Q&A] [answered]
Firstly, Prisma sounds great, its interesting to see the number of things it can do. I have gone through the docs, and have even done a few starter examples. I am using Prisma 2. 

The docs talk about two approaches to generate the Prisma Client --
1. Creating a database and introspecting it, and then refining your Prisma schema. 
2. Writing the Prisma Schema and then migrating it to the Database. (which is experimental, and I would not be wanting to use it at the moment). 

However, I cannot find docs that would allow me to see the underlying structure of the database tables. There is confusion that exists at present. 

Consider the following ER Diagram:
![Entity-Relationship Diagram](https://user-images.githubusercontent.com/29155477/82675181-1322d780-9c62-11ea-8bbd-854a0abd372d.png)

There are things like, the problem with above case I experience in general is 
1. **What would be the best approach to create these tables & Prisma schema models?** 
Should I create two disjoint `User` & `Event` tables? such that `EventAttendees` is not at all present. From what I can think of this can be done by adding an `attendee` field with reference to User. But, since an event may have multiple attendees, so that may make upto n rows of same event data. 
Hence, I feel the need to normalize this further into a different table. 
The question that further comes up is that if this normalization be worth it? Say, it'd been SQL. I may have had certainty to some point. But, I just started learning Prisma a few days ago. I am not certainly sure.

Now, coming further implementing it as such. I did encounter two problems while introspecting my database --

Q1. The `EventAttendee` database must have some unique identifier. I found two solutions, First to create an id field, second to use `@map`. Which would be better?  

Q2. The Prisma data models a type called Lists, represent as List L[], yet I am not sure how would I represent this in the database. Will the provided table model work? Yes, then how? or will I have to go with disjont model in discussed above. 

If you want to have a look at code for my generated Tables & Prisma Schema models. Please let me know.

**Comment by ryands17** [comment, marked as answer]
Hey @pavittarx :wave: 
Yes your're correct that there are two ways to go about this.

1. Using Prisma Migrate (experimental so not recommended to be used in production)

If you're going by the above approach, then creating the schema is easy as follows:

```prisma
model User {
  id             Int     @default(autoincrement()) @id
  name           String
  email          String  @unique
  mobile         String
  profile        String?
  attendedEvents Event[] @relation(references: [id])
  hostedEvent    Event[] @relation("host")
}

model Event {
  id        Int      @default(autoincrement()) @id
  title     String
  content   String
  location  String
  published Boolean
  attendees User[]   @relation(references: [id])
  host      User     @relation("host", fields: [hostId], references: [id])
  hostId    Int
  createdAt DateTime @default(now())
}
```

You are not required to create a join table as Migrate automatically does that for you.

2. Using Introspect (the recommended way in production)

For this, you need to create the main tables and a join table as follows:

```sql
create table "User" (
	id serial primary key,
	name text not null,
	email text not null,
	mobile text not null,
	profile text
);

create table "Event" (
	id serial primary key,
	title text,
	content text,
	location text,
	published boolean,
	createdAt timestamp with time zone default current_timestamp,
	hostId int,
	foreign key (hostId) references "User"(id)
);

create table "EventAttendees" (
	eventId integer NOT NULL,
	userId integer NOT NULL,
	primary key (eventId, userId),
	foreign key (eventId) references "Event" (id)
		on update cascade on delete cascade,
	foreign key (userId) references "User" (id)
		on update cascade on delete cascade
);
```

And then run `prisma introspect`, which will generate the relations for you.
You will notice I have used a composite key comprising of `eventId` and `userId` for the join table.

Let me know if that answers your query.

**Reply by pavittarx** [reply]
Indeed, approach 2 at this moment felt optimal, so I tried something similar. I decided to create independednt id column for "EventAttendees", however you example is still better. Thanks for your time to write it up. 

Prisma Introspect works really well, except only the fields are not in order the way they are defined in the table, they are correct but out of order. I was amazed, when after adding the `id` column in "EventAttendees" generated a schema I was looking for. 

Thanks.  <3

**Reply by ryands17** [reply]
@pavittarx 
You can also create implicit M-N relations where the third table is not required

For e.g.

1. Create the SQL in the following manner:

```sql
create table "User" (
	id serial primary key,
	name text not null,
	email text not null,
	mobile text not null,
	profile text
);

create table "Event" (
	id serial primary key,
	title text,
	content text,
	location text,
	published boolean,
	createdAt timestamp with time zone default current_timestamp,
	hostId int,
	foreign key (hostId) references "User"(id)
);

create table "_EventToUser" (
	"A" integer NOT NULL references "Event" (id),
	"B" integer NOT NULL references "User" (id)
);

create unique index "_EventToUser_AB_unique" ON "_EventToUser"("A", "B");

create index "_EventToUser_B_index" on "_EventToUser" ("B");
```

This follows the conventions for naming tables as described [here](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#conventions-for-relation-tables-in-implicit-m-n-relations).

2. Run `prisma introspect`. So you will get the following schema:

```prisma
model Event {
  content          String?
  createdat        DateTime? @default(now())
  hostid           Int?
  id               Int       @default(autoincrement()) @id
  location         String?
  published        Boolean?
  title            String?
  User_EventToUser User?     @relation(fields: [hostid], references: [id])
  User_EventToUser User[]    @relation(references: [id])
}

model User {
  email             String
  id                Int     @default(autoincrement()) @id
  mobile            String
  name              String
  profile           String?
  Event_EventToUser Event[]
  Event_EventToUser Event[] @relation(references: [id])
}
```

As this has generated similar field names for one-to-many and many-to-many relations, we need to change the names of those fields.

3. Update the schema:

```prisma
model Event {
  content          String?
  createdat        DateTime? @default(now())
  hostid           Int?
  id               Int       @default(autoincrement()) @id
  location         String?
  published        Boolean?
  title            String?
  host             User?     @relation(fields: [hostid], references: [id])
  User_EventToUser User[]    @relation("Attendees", references: [id])
}

model User {
  email             String
  id                Int     @default(autoincrement()) @id
  mobile            String
  name              String
  profile           String?
  hostedEvents      Event[]
  Event_EventToUser Event[] @relation("Attendees", references: [id])
}
```

So this is the schema without the many-to-many reference table.

**Reply by pavittarx** [reply]
Thank you @ryands17 for all this information.

**Comment by jeetmehta** [comment]
@pavittarx Thanks for this response. Just a quick question since your answer was over 1.5 years ago - is using Prisma Migrate now the recommended approach for production? Aka do we still need join tables for production systems / is that still the blessed path?

**Comment by ludralph** [comment]
Hi @pavittarx,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
